### ISSUE TITLE
NullPointerException

### ISSUE THREAD

**Issue by jacobembree** [hydrargyrum/epistolaire#1]
Installed using Fdroid version 0.3.4.  
When opening Epistolaire it crashes with the following text on the screen:

> SMS and MMS backup app  
> --------------------
> Please wait, backup is in progress.  It can take a while if you have a lot of SMSes or MMSes.  
Encountered an error :( java.lang.NullPointerException:  
Attempt to invoke virtual method 'boolean java.lang.String.equals(java.lang.Object)' on a null object reference

Here is the relevant output from `adb logcat`.

> 03-11 19:53:26.471  2411  2488 I WindowManager:   SURFACE show Surface(name=Splash Screen re.indigo.epistolaire)/@0x6ef495c: Splash Screen re.indigo.epistolaire
03-11 19:53:26.554  3079 20735 E DatabaseUtils: Writing exception to parcel
03-11 19:53:26.554  3079 20735 E DatabaseUtils: java.lang.NullPointerException: Attempt to invoke virtual method 'boolean java.lang.String.equals(java.lang.Object)' on a null object reference
03-11 19:53:26.554  3079 20735 E DatabaseUtils: 	at android.database.sqlite.SQLiteQueryBuilder.buildUnionSubQuery(SQLiteQueryBuilder.java:696)
03-11 19:53:26.554  3079 20735 E DatabaseUtils: 	at com.android.providers.telephony.MmsSmsProvider.getConversations(MmsSmsProvider.java:918)
03-11 19:53:26.554  3079 20735 E DatabaseUtils: 	at com.android.providers.telephony.MmsSmsProvider.query(MmsSmsProvider.java:394)
03-11 19:53:26.554  3079 20735 E DatabaseUtils: 	at android.content.ContentProvider.query(ContentProvider.java:1057)
03-11 19:53:26.554  3079 20735 E DatabaseUtils: 	at android.content.ContentProvider.query(ContentProvider.java:1149)
03-11 19:53:26.554  3079 20735 E DatabaseUtils: 	at android.content.ContentProvider$Transport.query(ContentProvider.java:241)
03-11 19:53:26.554  3079 20735 E DatabaseUtils: 	at android.content.ContentProviderNative.onTransact(ContentProviderNative.java:102)
03-11 19:53:26.554  3079 20735 E DatabaseUtils: 	at android.os.Binder.execTransact(Binder.java:746)
03-11 19:53:26.577  2411 29823 V WindowManager: Focus changed from null to Window{25a853a u0 re.indigo.epistolaire/re.indigo.epistolaire.MainActivity}
03-11 19:53:26.629  2411  2488 I WindowManager: Window drawn AppWindowToken{9d3b7fb token=Token{545a58a ActivityRecord{ca1a3f5 u0 re.indigo.epistolaire/.MainActivity t4014}}}
03-11 19:53:26.630  2411  2488 I WindowManager:   SURFACE show Surface(name=re.indigo.epistolaire/re.indigo.epistolaire.MainActivity)/@0x9fe95eb: re.indigo.epistolaire/re.indigo.epistolaire.MainActivity
03-11 19:53:26.655  2411  2439 I LaunchCheckinHandler: Displayed re.indigo.epistolaire/.MainActivity,wp,ca,240
03-11 19:53:26.672  2411  2439 I ActivityManager: Displayed re.indigo.epistolaire/.MainActivity: +214ms
03-11 19:53:26.816   891  2199 W SurfaceFlinger: Attempting to set client state on removed layer: Splash Screen re.indigo.epistolaire#0
03-11 19:53:26.816   891  2199 W SurfaceFlinger: Attempting to destroy on removed layer: Splash Screen re.indigo.epistolaire#0

**Comment by hydrargyrum** [comment]
Hmm, the stacktrace from logcat is not very informative.
I will try to use a framework like ACRA (open source crashlytics, usable on fdroid).

In the meantime, can you provide more info? What android version are you using? Do you use the stock SMS app?

**Comment by hydrargyrum** [comment]
I just released epistolaire 0.4.0 which:
- skips some non-essential errors
- adds more logging, especially with full stacktraces to logcat (in a future release, I will use [ACRA](https://github.com/ACRA/acra/) as a crash report tool)

It might not fix the error you have, but at least we'll have more info.

Now, f-droid takes a few days to scan repositories and publish new versions, so wait until 0.4.0 hits f-droid.

**Comment by hydrargyrum** [comment]
@jacobembree, 0.4.0 is now available on f-droid, could you retry if your issue is fixed, or at least if the logcat is more verbose?

**Comment by jacobembree** [comment]
This is with 0.4.0.  Does this help?

```
03-22 12:50:33.457  2411 13392 I ActivityManager: START u0 {act=android.intent.action.MAIN cat=[android.intent.category.LAUNCHER] flg=0x10200000 cmp=re.indigo.epistolaire/.MainActivity bnds=[426,982][559,1164]} from uid 10109
03-22 12:50:33.494   871   889 E ANDR-PERF-RESOURCEQS: Failed to apply optimization [4, 0]
03-22 12:50:33.518 31812 31812 W ActivityThread: handleWindowVisibility: no activity for token android.os.BinderProxy@137fa95
03-22 12:50:33.528  2411 13392 V WindowManager: Focus changed from Window{ff45bef u0 com.motorola.launcher3/com.android.launcher3.CustomizationPanelLauncher} to null
03-22 12:50:33.572  2411  2488 I WindowManager:   SURFACE show Surface(name=Splash Screen re.indigo.epistolaire)/@0x8283aac: Splash Screen re.indigo.epistolaire
03-22 12:50:33.753  3079  3261 E DatabaseUtils: Writing exception to parcel
03-22 12:50:33.753  3079  3261 E DatabaseUtils: java.lang.NullPointerException: Attempt to invoke virtual method 'boolean java.lang.String.equals(java.lang.Object)' on a null object reference
03-22 12:50:33.753  3079  3261 E DatabaseUtils: 	at android.database.sqlite.SQLiteQueryBuilder.buildUnionSubQuery(SQLiteQueryBuilder.java:696)
03-22 12:50:33.753  3079  3261 E DatabaseUtils: 	at com.android.providers.telephony.MmsSmsProvider.getConversations(MmsSmsProvider.java:918)
03-22 12:50:33.753  3079  3261 E DatabaseUtils: 	at com.android.providers.telephony.MmsSmsProvider.query(MmsSmsProvider.java:394)
03-22 12:50:33.753  3079  3261 E DatabaseUtils: 	at android.content.ContentProvider.query(ContentProvider.java:1057)
03-22 12:50:33.753  3079  3261 E DatabaseUtils: 	at android.content.ContentProvider.query(ContentProvider.java:1149)
03-22 12:50:33.753  3079  3261 E DatabaseUtils: 	at android.content.ContentProvider$Transport.query(ContentProvider.java:241)
03-22 12:50:33.753  3079  3261 E DatabaseUtils: 	at android.content.ContentProviderNative.onTransact(ContentProviderNative.java:102)
03-22 12:50:33.753  3079  3261 E DatabaseUtils: 	at android.os.Binder.execTransact(Binder.java:746)
03-22 12:50:33.762 31812 19377 E MainActivity: crash when dumping messages database
03-22 12:50:33.762 31812 19377 E MainActivity: java.lang.NullPointerException: Attempt to invoke virtual method 'boolean java.lang.String.equals(java.lang.Object)' on a null object reference
03-22 12:50:33.762 31812 19377 E MainActivity: 	at android.os.Parcel.createException(Parcel.java:1956)
03-22 12:50:33.762 31812 19377 E MainActivity: 	at android.os.Parcel.readException(Parcel.java:1918)
03-22 12:50:33.762 31812 19377 E MainActivity: 	at android.database.DatabaseUtils.readExceptionFromParcel(DatabaseUtils.java:183)
03-22 12:50:33.762 31812 19377 E MainActivity: 	at android.database.DatabaseUtils.readExceptionFromParcel(DatabaseUtils.java:135)
03-22 12:50:33.762 31812 19377 E MainActivity: 	at android.content.ContentProviderProxy.query(ContentProviderNative.java:418)
03-22 12:50:33.762 31812 19377 E MainActivity: 	at android.content.ContentResolver.query(ContentResolver.java:802)
03-22 12:50:33.762 31812 19377 E MainActivity: 	at android.content.ContentResolver.query(ContentResolver.java:752)
03-22 12:50:33.762 31812 19377 E MainActivity: 	at android.content.ContentResolver.query(ContentResolver.java:710)
03-22 12:50:33.762 31812 19377 E MainActivity: 	at re.indigo.epistolaire.MmsDumper.allMessages(MmsDumper.kt:34)
03-22 12:50:33.762 31812 19377 E MainActivity: 	at re.indigo.epistolaire.MmsDumper.getJson(MmsDumper.kt:27)
03-22 12:50:33.762 31812 19377 E MainActivity: 	at re.indigo.epistolaire.MainActivity$DumpTask.doInBackground(MainActivity.kt:81)
03-22 12:50:33.762 31812 19377 E MainActivity: 	at re.indigo.epistolaire.MainActivity$DumpTask.doInBackground(MainActivity.kt:75)
03-22 12:50:33.762 31812 19377 E MainActivity: 	at android.os.AsyncTask$2.call(AsyncTask.java:333)
03-22 12:50:33.762 31812 19377 E MainActivity: 	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
03-22 12:50:33.762 31812 19377 E MainActivity: 	at android.os.AsyncTask$SerialExecutor$1.run(AsyncTask.java:245)
03-22 12:50:33.762 31812 19377 E MainActivity: 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
03-22 12:50:33.762 31812 19377 E MainActivity: 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
03-22 12:50:33.762 31812 19377 E MainActivity: 	at java.lang.Thread.run(Thread.java:798)
03-22 12:50:33.768  2411 13392 V WindowManager: Focus changed from null to Window{87f9f0a u0 re.indigo.epistolaire/re.indigo.epistolaire.MainActivity}
03-22 12:50:33.865  2411  2488 I WindowManager: Window drawn AppWindowToken{4ab918b token=Token{a35ab5a ActivityRecord{7f9df05 u0 re.indigo.epistolaire/.MainActivity t4383}}}
03-22 12:50:33.865  2411  2488 I WindowManager:   SURFACE show Surface(name=re.indigo.epistolaire/re.indigo.epistolaire.MainActivity)/@0xc87f37b: re.indigo.epistolaire/re.indigo.epistolaire.MainActivity
03-22 12:50:33.876  2411  2439 I LaunchCheckinHandler: Displayed re.indigo.epistolaire/.MainActivity,wp,ca,361
03-22 12:50:33.876  2411  2439 I ActivityManager: Displayed re.indigo.epistolaire/.MainActivity: +350ms
03-22 12:50:33.964  3594  3594 D LauncherAppWidgetHost: setListenIfResumed :false, mFlags:5
03-22 12:50:33.964  3594  3594 D LauncherAppWidgetHost: stopListening
03-22 12:50:33.965  3594  3594 W RecentsModel: onTrimMemory level = 20
03-22 12:50:34.059   891  1478 W SurfaceFlinger: Attempting to set client state on removed layer: Splash Screen re.indigo.epistolaire#0
03-22 12:50:34.059   891  1478 W SurfaceFlinger: Attempting to destroy on removed layer: Splash Screen re.indigo.epistolaire#0
```

**Comment by hydrargyrum** [comment]
@jacobembree, thanks, I have [resolved](https://github.com/hydrargyrum/epistolaire/commit/fb80a30406aa7bfb2022ef0a17c139b425757ad9) yesterday an issue having the same backtrace as yours and subsequently released 0.4.1, which will take (again) a few days to hit f-droid. Hoping this time your problem will be nailed!

**Comment by jacobembree** [comment]
My friend installed the update and reports success.  Thanks @hydrargyrum!
Relevant commit for other readers: fb80a30406aa7bfb2022ef0a17c139b425757ad9
